### PR TITLE
[FEATURE, PARAM] 13C isotope scoring for FF-Metabo

### LINKED ISSUE #2016: FeatureFinderMetabo performs isotopic deconvolution incorrectly
Often, FFM doesn't merge mass traces of low-medium intensity into features. As discussed privately with @mrurik the problem lies in the strict matching of mass traces with isotopic model that allows 3 or 5% deviation. @mrurik proposed two possible ways of fixing it:

1. option to disable matching with isotopic model
2. option to use a new model with 10% error

### PR DESCRIPTION
**Pull request by cbielow**
use 13C mass trace model for isotope assembly instead of the model of Kenar et al (both modes are still available; 13C is default).

On a lipidomics datasets, the new model produces slightly less features in total (which is good, because more traces are assembled)
Input traces:    4184
FF-legacy isotope model features: 2703
FF-13C isotope model features: 2659 

Should fix #2016 (FeatureFinderMetabo performs isotopic deconvolution incorrectly).

Looking at the mass differences, 13C is the correct model:

image

 - green dashed line is the median -- fits theoretical 13C (solid green line) very well
 - magenta lines are the shifts from the legacy model (no good fit at all -- the data is 99% charge 1, so should be on the lowest line according to the legacy model)
 - grey lines are other mass defects (hydrogen, oxygen etc... also no fit)
 - points are colored by isotope position (all seem to obey 13C)
 - less intense points have higher discrepencies -- very much expected due to inferior S/N
 - the plot is not biased by the model used -- both models produce roughly the same plot

Other minor improvements:

 - New parameter 'report_convex_hulls' yielding a size reduction of about 1:10
     - report_convex_hulls::true  --> 23MB
     - report_convex_hulls::false -->  2MB
 - moved function implementations to .C file
 - removed dead code
 - TOPPView: show 4th digit after dot of isotope distances in 1D view -- to check results more conveniently (i.e. 13C is 1.003355)

### CODE REVIEW

**Comment by cbielow** [comment]
I also checked singleton mass traces after FF-Metabo but I could not find false-negative traces (ie., which should have been assembled but in fact were not). But my settings for finding mass traces are pretty stringent (i.e. the default settings) -- so this might be a problem with VERY low intensity traces with bad support, but that would need checking. Its definitely not a problem all traces I can see.

**Comment by cbielow** [comment]
TOPPAS workflow and plotting script -- if someone wants to try on other data

[isotope_TOPPAS and plot.zip](https://github.com/OpenMS/OpenMS/files/401727/isotope_TOPPAS.and.plot.zip)

**Comment by timosachsenberg** [comment]
nice addition. Fits well to my speculations ;)

**Comment by iprotsyuk** [comment]
@cbielow, could you please have a look at these two files: [1](https://gist.github.com/iprotsyuk/99d9a8ef7000dcff889caf8793cfba39), [2](https://gist.github.com/iprotsyuk/58129a6b189a22f88fb19dbc5f67778e). Before, I found false-negatives there, asked advice from @mrurik, and he suggested it's the isotopic model issue.

What we saw, for example, in the first file is a mass trace around mz 523.286, RT 200 s that have satellite traces located 0.5 and 1 Dalton above, but FeatureFinderMetabo interprets them as different features. However, the second file is another sample from the same study, and it contains same mass traces with higher intensities, and they’re joined into a single feature.

Maybe there's a combination of FF-Metabo parameters that can solve this, but we didn't manage to find it.

**Comment by cbielow** [comment]
well... in the first file, the 'feature' that we're looking at does not even have a proper isotope distribution (M+1 is about 2x higher than M+0). This thing is well in the noise and no isotope model is going to rescue that. You'd need to switch isotope filtering off in order to call that feature. This is certainly possible, and one can think of a model which gradually degrades its restrictions as the intensity of features drops -- basically allowing anything with the correct m/z distance and ignoring intensity at the noise-level - but that would require some thought.

In file 2, intensities become more decent (35k), so you can identify the feature. Isotopes look decent as well.

I'm not sure if its worth investing a lot of time here, since you will always find cases like this where the call is borderline, even for a human expert.

**Comment by iprotsyuk** [comment]
@cbielow agree, that wasn't a good example. What about mz 1000.544, RT 205 in the 1st file? Isotopic pattern seems ok, but still 3 mass traces with 1 Th difference aren't joined. Does it mean intensities are a bit off the isotopic model?

Agree with you regarding the implementation of a flexible model working for different intensities. It doesn't seem to be a crucial feature.

**Comment by cbielow** [comment]
indeed, this feature is found, but the isotope filtering discards it again.
So for now, the only option I can see is: disable the filter.

**Comment by iprotsyuk** [comment]
@cbielow, will this new option be implemented within this issue? Otherwise, I can submit a new one.

**Comment by cbielow** [comment]
Do you mean disabling the isotope filter?  Thats already possible,  see advanced parameters. 
-- 
Sent via mobile phone.

Am 5. August 2016 20:43:12 MESZ, schrieb Ivan Protsyuk <notifications@github.com>:

**Comment by hroest** [comment]
@cbielow is this mostly adopted for lipids (which are a lot of Cs) or is this generally better for all metabolomics? If its only better for lipids, then we maybe shouldnt change the default but if it improves for all metabolomics datasets, we should change it.

**Comment by iprotsyuk** [comment]
@cbielow, found it. Thanks!

**Comment by cbielow** [comment]
I do not have data for general metabolomics available, so I cant say.
However, for peptides, both models seem "suboptimal"

image

The green line is 13C. The grey lines are other elements (N, O,....), and the thick yellow, blue and red lines are for isotope 1,2,3 (from the old model). The `dashed` counterparts are what the data says... and thats actually divergent, i.e. the trend goes the opposite way.

**Comment by hroest** [comment]
well that is why I am wondering whether we should make 13C the default, especially as the other model is published. 

Isnt this why we have averagine for peptides? I guess that would be better for peptides?

**Comment by timosachsenberg** [comment]
@cbielow can you do the same plot for the i.e. top 10% intensity quantile of monoisotopic traces?

**Comment by FabianAicheler** [comment]
Maybe this is useful: https://gist.github.com/FabianAicheler/e8a4106fb7d6821199057fe707b24224
Contains mzML of the CAMERA dataset, in this case ~34 spike-in metabolites (some amino acids included) without background, and FFM with the original model I did with annotations of the compounds (according to their standards list)

Also, the CAMERA data at http://www.ebi.ac.uk/metabolights/MTBLS188 contains these standards in different concentrations with a leaf extract background, which I think is mainly metabolomics-based (methanolic extract, see metabolights protocols for details)?. Maybe you could use that for distance distribtutions.

**Comment by cbielow** [comment]
@timosachsenberg: the plot does not change for peptides

**Comment by cbielow** [comment]
@FabianAicheler Using your dataset, a third scenario emerges
image

However, if I take another (random) dataset from the CAMERA repo (Mspos_Leaf_MM48_200nM_1-A,3_01_14601.mzData), I get yet another result
image

Note that it does not matter what distance scoring I use (13C or old) -- neither for the graph, nor for the number of features.

So IMHO we can draw two conclusions:

1) if we are looking at LC-MS data, the model does not matter (since there is little chance that isotope traces will interfere). The assembled features are 98% identical.
2) for shotgun lipidomics, trace interference is a big problem and 13C must be used.
3) the predicted isotope distances from the old model are usually not reflecting the observed data (iso0 and iso1 distances are usually swapped) -- but this is a very limited set of data we are looking at here. There could be other MTX data, which reflect the model well.

So what should be the default?

**Comment by cbielow** [comment]
@hroest The averagine for peptides predicts the intensities, not the m/z. And the composition is fixed, i.e. the relative contributions within one isotope peak remain constant. -- or am I missing the point here?

**Comment by timosachsenberg** [comment]
"The assembled features are 98% identical."
ok but what about the number of assembled vs. single mass traces?

**Comment by cbielow** [comment]
> "The assembled features are 98% identical."
> ok but what about the number of assembled vs. single mass traces?

that's what I meant. There is no difference worth mentioning.

**Comment by cbielow** [comment]
on another note: We could simplify the parameters of the ffm section a lot, i.e.

Old:
 - ffm:disable_isotope_filtering (true,false)
 - ffm:isotope_model (metabolites, peptides)
 - ffm:isotope_noisemodel (5%, 10%)

To:
 - ffm:isotope_model (metabolites, none)
 - ffm:isotope_noisemodel (5%, 10%)

or even more drastically
 - ffm:isotope_model (metabolites (%5 noise), metabolites (10% noise), none)

because ATM, `peptides` effectively means: no isotope filtering and is IMHO highly misleading

**Comment by timosachsenberg** [comment]
oh ok makes sense. I thought at least averagines are used.

**Comment by cbielow** [comment]
it uses averagines actually.. but at a completely different position. But it does not change the issue stated above (the peptide model was used irrespective of what `ffm:disable_isotope_filtering` was set to. Fixed now.

**Comment by cbielow** [comment]
Since the results on general metabolomics data are not favouring any model in particular, I've made the old model the default again.
The set of parameters is now minimized and reflects internal workings much better.

Any more comments?!
Otherwise I'd merge.

Failed tests are unrelated

File: src/topp/FeatureFinderMetabo.cpp [hunk 1]
```diff
@@ -76,12 +76,24 @@ using namespace std;
...
-  hypotheses are formulated and scored according to how well differences in RT
-  and m/z or intensity ratios match to those of theoretical isotope patterns.
+  hypotheses are formulated and scored according to how well differences in RT (optional),
+  m/z or intensity ratios match to those of theoretical isotope patterns.
 
   If the raw data scans contain the scan polarity information, it is stored as
   meta value "scan_polarity" in the output file.
 
+  Mass trace clustering can be done using either 13C distances or a linear model (Kenar et al) -- see parameter 'ffm:mz_scoring_13C'.
+  Generally, for lipidomics, use 13C, since lipids contain a lot of 13C.
```
Comment (hroest) [inline src/topp/FeatureFinderMetabo.cpp]: :+1:

**Comment by timosachsenberg** [comment]
> @hroest The averagine for peptides predicts the intensities, not the m/z. And the composition is fixed, i.e. the relative contributions within one isotope peak remain constant. -- or am I missing the point here?

I think he means the averagine model could also be used to estimate the m/z spacing (i.e. as isotope ratio weighted average of m/z spacing). The isotopic abundance of each element (with two isotopes) can be calculated from a binomial distribution scaled by the elements isotopic abundance. So at isotopic position x you could calculate the expected spacing by averaging each elements m/z spacing (iso 1 - iso 0) weighted by its x-th isotopic abundance. This could e.g. be precalculated.

**Comment by cbielow** [comment]
This could indeed work, but I think a simple weighted average is not the 
way to go, since we are looking at overlapping Gaussians here, i.e. the 
apex of the observed mixture is not neccesarily governed by the average 
of the underlying Gaussian centers. For two Gaussians, this should look 
more like a sigmoidal, i.e. a small vs. a big Gaussian, the big one's 
apex will determine the apex of the mixture (which is not the weighted 
average of the two).

Am 8/11/2016 um 10:36 AM schrieb Timo Sachsenberg:

**Comment by timosachsenberg** [comment]
Yeah it's just crude approximation and doesn't take the peak width into account. For the stable isotope probing I once wrote some code that does exact isotope pattern calculations, convolution with a gaussian and subsequent peak picking to simulate the effect of peak width on the resolution of isotopic peaks. For SIP peptides it had no benefits on the isotopic trace detection and was much slower so I removed the code. For SIP peptides (and also for peptides with natural isotopic composition) and realistic (aka standard orbitrap MS1) resolutions I got either a single apex (basically the N15 and C13 peaks melt to a wider gaussian), this sigmoidal shape that leads to one or two peaks or a clear separation of both peaks.

**Pull request merged**